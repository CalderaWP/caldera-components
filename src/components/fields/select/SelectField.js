import React from 'react';
import {fieldPropTypes} from '../propTypes';

/**
 * Creates a select field
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export const SelectField = (props) => {
	/**
	 * Dispatches new value(s) to parent
	 *
	 * @param selection
	 * @returns {*}
	 */
	function changeHandler(selection) {
		return props.onValueChange(selection);
	}

	return (
		<select
			id={props.id}
			className={props.className}
			value={props.value}
			onChange={changeHandler}
		>
			{props.options
				.map((item) => {
					return(
						<option
							key={item.value}
							className={ 'caldera-config-option' }
							value={item.value}
						>
							{item.label}
						</option>
					);
				})
			}
		</select>



	);
};

/**
 * Prop definition for select fields
 *
 * @type {{}}
 */
SelectField.propTypes = {
	...fieldPropTypes,
};

/**
 * Default props for select fields
 *
 * @type {{options: Array}}
 */
SelectField.defaultProps = {
	options: []
};