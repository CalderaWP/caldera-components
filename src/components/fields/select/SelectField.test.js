import {SelectField} from './SelectField';
import renderer from 'react-test-renderer';
import React from 'react';
import { mount } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

describe( 'Select field component', () => {
	function genericChangeHandler() {

	}

	describe( 'Props', () => {
		it( 'Creates without options or value', () => {
			const component = renderer.create(
				<SelectField
					id={'r'}
					fieldClassName={'rs'}
					onValueChange={genericChangeHandler}
				/>
			);
			expect( component.toJSON() ).toMatchSnapshot();
		});

		it( 'Creates with options', () => {
			const component = renderer.create(
				<SelectField
					id={'r'}
					fieldClassName={'rs'}
					onValueChange={genericChangeHandler}
					options={[
						{
							value: 1,
							label: 'One'
						}
					]}
				/>
			);
			expect( component.toJSON() ).toMatchSnapshot();
		});

		it( 'Creates with options and value', () => {
			const component = renderer.create(
				<SelectField
					id={'r'}
					fieldClassName={'rs'}
					onValueChange={genericChangeHandler}
					options={[
						{
							value: '1',
							label: 'One'
						}
					]}
					value={'1'}
				/>
			);
			expect( component.toJSON() ).toMatchSnapshot();
		});

		it( 'sets id attr', () => {
			const wrapper = mount(
				<SelectField
					id={'r3s'}
					fieldClassName={'rs'}
					onValueChange={genericChangeHandler}
				/>
			);
			expect(wrapper.find('select').prop('id')).toBe('r3s');
		});

		describe( 'Options ', () => {
			it( 'Has option', () =>{
				const wrapper = mount(
					<SelectField
						id={'r'}
						fieldClassName={'rs'}
						onValueChange={genericChangeHandler}
						options={[
							{
								value: 1,
								label: 'One'
							},
							{
								value: 2,
								label: 'Two'
							}
						]}
					/>
				);
				expect(wrapper.find('.caldera-config-option')).toHaveLength(2);
			});



		});

		describe( 'Value of field', () => {
			it( 'Set value from default', () => {
				const wrapper = mount(
					<SelectField
						id={'r'}
						fieldClassName={'rs'}
						onValueChange={() => {}}
						options={[
							{
								value: 'one',
								label: 'One'
							},
							{
								value: '2',
								label: 'Two'
							}
						]}
						value={'2'}
					/>
				);
				expect( wrapper.find('select').prop( 'value' ) ).toEqual('2');
			});
		});

	});




});