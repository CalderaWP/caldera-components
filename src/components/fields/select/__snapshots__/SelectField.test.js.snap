// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select field component Props Creates with options 1`] = `
<select
  className={undefined}
  id="r"
  onChange={[Function]}
  value={undefined}
>
  <option
    className="caldera-config-option"
    value={1}
  >
    One
  </option>
</select>
`;

exports[`Select field component Props Creates with options and value 1`] = `
<select
  className={undefined}
  id="r"
  onChange={[Function]}
  value="1"
>
  <option
    className="caldera-config-option"
    value="1"
  >
    One
  </option>
</select>
`;

exports[`Select field component Props Creates without options or value 1`] = `
<select
  className={undefined}
  id="r"
  onChange={[Function]}
  value={undefined}
/>
`;
